### PR TITLE
Write min_log_number_to_keep to MANIFEST during atomic flush under 2 phase commit

### PR DESCRIPTION
**Pull request by cheng-chang**
When 2 phase commit is enabled, if there are prepared data in a WAL, the WAL should be kept, the minimum log number for such a WAL is written to MANIFEST during flush. In atomic flush, such information is not written to MANIFEST.

Test Plan:
Added a new unit test `DBAtomicFlushTest.ManualFlushUnder2PC`, this test fails in atomic flush without this PR, after this PR, it succeeds.

### CODE REVIEW

File: db/memtable_list.cc [hunk 1]
```diff
@@ -726,6 +727,17 @@ Status InstallMemtableAtomicFlushResults(
     assert(0 == num_entries);
   }
 
+  if (vset->db_options()->allow_2pc) {
+    uint64_t min_log_number_to_keep = port::kMaxUint64;
```
Comment (riversand963) [inline db/memtable_list.cc:731]: Nit: `port::kMaxUint64`.

File: db/memtable_list.cc [hunk 2]
```diff
@@ -726,6 +727,17 @@ Status InstallMemtableAtomicFlushResults(
     assert(0 == num_entries);
   }
 
+  if (vset->db_options()->allow_2pc) {
```
Comment (riversand963) [inline db/memtable_list.cc:730]: Should the non-2pc case be handled too?

**Review by riversand963 — COMMENTED** [review]
Thanks @cheng-chang for the fix. Left a few comments.

File: db/memtable_list.cc (hunk 2 above)
Comment (cheng-chang) [reply db/memtable_list.cc:730]: The algorithm for computing the min_log_number has problems, the non-2pc case is fixed in https://github.com/facebook/rocksdb/pull/7660. I'll update this PR after https://github.com/facebook/rocksdb/pull/7660 is merged.

File: db/memtable_list.cc (hunk 1 above)
Comment (cheng-chang) [reply db/memtable_list.cc:731]: hmm, already used it

**Comment by cheng-chang** [comment]
@riversand963 PTAL, thanks.

File: HISTORY.md [hunk 3]
```diff
@@ -21,6 +21,7 @@
 * Fixed a potential bug caused by evaluating `TableBuilder::NeedCompact()` before `TableBuilder::Finish()` in compaction job. For example, the `NeedCompact()` method of `CompactOnDeletionCollector` returned by built-in `CompactOnDeletionCollectorFactory` requires `BlockBasedTable::Finish()` to return the correct result. The bug can cause a compaction-generated file not to be marked for future compaction based on deletion ratio.
 * Fixed a seek issue with prefix extractor and timestamp.
 * Fixed a bug of encoding and parsing BlockBasedTableOptions::read_amp_bytes_per_bit as a 64-bit integer.
+* Fixed a bug of a recovery corner case, details in PR7621.
```
Comment (riversand963) [inline HISTORY.md:24]: Is this PR related to this item?

Comment (cheng-chang) [reply HISTORY.md:24]: committed by mistake, removed

File: db/memtable_list.cc [hunk 4]
```diff
@@ -752,14 +746,21 @@ Status InstallMemtableAtomicFlushResults(
...
-        PrecomputeMinLogNumberToKeepNon2PC(vset, cfds, edit_lists);
-    const auto& wals = vset->GetWalSet().GetWals();
-    if (!wals.empty() && min_wal_number_to_keep > wals.begin()->first) {
+  if (vset->db_options()->track_and_verify_wals_in_manifest &&
+      !vset->GetWalSet().GetWals().empty()) {
+    if (!vset->db_options()->allow_2pc) {
+      min_wal_number_to_keep =
+          PrecomputeMinLogNumberToKeepNon2PC(vset, cfds, edit_lists);
+    }
+    if (min_wal_number_to_keep > vset->GetWalSet().GetWals().begin()->first) {
```
Comment (riversand963) [inline db/memtable_list.cc:763]: Can we do something like:
```
WalNumber min_wal_number_to_keep = allow_2pc ? PrecomputeMinLogNumberToKeep2PC() : PrecomputeMinLogNumberToKeepNon2PC()?
edit.SetMinLogNumberToKeep(min_wal_ number_to_keep);
```
I guess I am not sure I miss anything when I see `min_wal_number_to_keep` is evaluated in allow_2pc mode but only conditionally in non-2pc mode. In either case, it would be OK for the db to track min_wal_number_to_keep, right?

File: db/memtable_list.cc [hunk 5]
```diff
@@ -674,20 +674,13 @@ void MemTableList::RemoveMemTablesOrRestoreFlags(
...
     // if the performance has some problem.
-    bool should_skip = false;
-    for (MemTable* m_to_flush : memtables_to_flush) {
-      if (m == m_to_flush) {
-        should_skip = true;
-        break;
-      }
-    }
-    if (should_skip) {
+    if (memtables_to_flush.count(m)) {
```
[the remark this answers is not in the document]
Comment (cheng-chang) [reply db/memtable_list.cc:683]: removed

File: db/memtable_list.cc (hunk 4 above)
Comment (cheng-chang) [reply db/memtable_list.cc:763]: I agree that it's a bit confusing. But according to https://github.com/facebook/rocksdb/blob/master/db/version_set.cc#L4254, the min_log_number_to_keep in VersionEdit now should only be set in 2pc mode.

File: db/column_family.cc [hunk 6]
```diff
@@ -708,9 +708,8 @@ uint64_t ColumnFamilyData::OldestLogToKeep() {
   auto current_log = GetLogNumber();
 
   if (allow_2pc_) {
-    autovector<MemTable*> empty_list;
-    auto imm_prep_log =
-        imm()->PrecomputeMinLogContainingPrepSection(empty_list);
+    std::unordered_set<MemTable*> empty_set;
```
[the remark this answers is not in the document]
Comment (cheng-chang) [reply db/column_family.cc:711]: We can use std::optional, but that requires c++17. I'll change the parameter to be a pointer, then passing nullptr to indicate that it's optional.

**Comment by cheng-chang** [comment]
@riversand963 thanks for comments, PTAL.

File: db/column_family.cc (hunk 6 above)
[the remark this answers is not in the document]
Comment (riversand963) [reply db/column_family.cc]: If you go with pointer type, then maybe add a comment to explain what `nullptr` means here.
I am also OK with keeping it `const std::unordered_set&`.

**Review by riversand963 — APPROVED** [review]

File: db/db_impl/db_impl_files.cc [hunk 7]
```diff
@@ -681,6 +683,37 @@ uint64_t FindMinPrepLogReferencedByMemTable(
   return min_log;
 }
 
+uint64_t FindMinPrepLogReferencedByMemTable(
+    VersionSet* vset, const autovector<ColumnFamilyData*>& cfds_to_flush,
+    const autovector<const autovector<MemTable*>*>& memtables_to_flush) {
+  uint64_t min_log = 0;
```
[the remark this answers is not in the document]
Comment (cheng-chang) [reply db/db_impl/db_impl_files.cc:689]: the current logic assumes the return value is 0 if no log contains prep log (since no log will use 0 as log number).
